lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,136 @@
+    HasAccess(user, login, node, role) <- HasAllowRole(user, login, node, role), !HasDenyRole(user, node, _), !HasDeniedLogin(user, login, _);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasTrait(user, LOGIN_TRAIT_HASH, login);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasAllowRole(user, login, node, _);
+
+    DenyLogins(user, login, role) <- HasDeniedLogin(user, login, role);
+}
+
+#[no_mangle]
+pub extern "C" fn process_access(input: *mut c_uchar, output: *mut c_uchar, input_len: size_t, output_len: size_t) -> i32 {
+    let mut runtime = Crepe::new();
+    let b = unsafe { slice::from_raw_parts_mut(input, input_len) };
+    let r = match types::Facts::decode(BytesMut::from(&b[..])) {
+        Ok(b) => b,
+        Err(_e) => return Errors::InvalidInputError as i32
+    };
+
+    for pred in &r.predicates {
+        if pred.name == types::facts::PredicateType::Hasrole as i32 {
+            runtime.extend(&[HasRole(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Hastrait as i32 {
+            runtime.extend(&[HasTrait(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Roleallowslogin as i32 {
+            runtime.extend(&[RoleAllowsLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Roledenieslogin as i32 {
+            runtime.extend(&[RoleDeniesLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Roleallowsnodelabel as i32 {
+            runtime.extend(&[RoleAllowsNodeLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Roledeniesnodelabel as i32 {
+            runtime.extend(&[RoleDeniesNodeLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Nodehaslabel as i32 {
+            runtime.extend(&[NodeHasLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else {
+            return Errors::InvalidInputError as i32;
+        }
+    }
+
+    let (accesses, deny_accesses, deny_logins) = runtime.run();
+    let mut predicates = vec![];
+    predicates.extend::<Vec<_>>(accesses.into_iter().map(|HasAccess(a, b, c, d)| types::facts::Predicate{
+        name: types::facts::PredicateType::Hasaccess as i32,
+        atoms: vec![a, b, c, d],
+    }).collect());
+    predicates.extend::<Vec<_>>(deny_accesses.into_iter().map(|DenyAccess(a, b, c, d)| types::facts::Predicate{
+        name: types::facts::PredicateType::Denyaccess as i32,
+        atoms: vec![a, b, c, d],
+    }).collect());
+    predicates.extend::<Vec<_>>(deny_logins.into_iter().map(|DenyLogins(a, b, c)| types::facts::Predicate{
+        name: types::facts::PredicateType::Denylogins as i32,
+        atoms: vec![a, b, c],
+    }).collect());
```

Comment:
`Vec::extend` comes from the `Extend` trait and takes anything that can be converted into an iterator. No need to collect into a `Vec` before extending.